### DISCUSSION TITLE
husky + lint-stagedの導入でcommitできない現象

### DISCUSSION THREAD

**Discussion by ghost** [lgtm-factory/lgtm-factory#86] [Q&A] [answered]
https://zenn.dev/risu729/articles/latest-husky-lint-staged
これのTL;DRでコマンド打った後にcommitしようとすると

```
git --no-optional-locks -c color.branch=false -c color.diff=false -c color.status=false -c diff.mnemonicprefix=false -c core.quotepath=false -c credential.helper=sourcetree commit -q -F /var/folders/kc/qkl2lph97x1g7n53185wpr440000gn/T/SourceTreeTemp.fQxhZJ -a 
Cannot use import statement outside a module
(node:8756) Warning: require() of ES modules is not supported.
require() of /Users/sorao/Desktop/lgtm-factory/node_modules/lint-staged/bin/lint-staged.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename lint-staged.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/sorao/Desktop/lgtm-factory/node_modules/lint-staged/package.json.
husky - pre-commit script failed (code 1)
Completed with errors, see above
```

のエラーが吐き出される。
GPTの解決策は以下。

```
1. lint-staged.js を .cjs ファイルにリネーム
ファイル拡張子を .cjs に変更することで、Node.js はそのファイルを CommonJS として扱います。この方法は簡単ですが、パッケージのバージョンアップによって再度問題が発生する可能性があります。

2. import() による動的インポートに変更
require() の代わりに、動的な import() 関数を使うようにコードを書き換える方法です。これはモダンな方法であり、ES モジュールとの互換性を保つことができます。

js
コードをコピーする
import('./lint-staged.js').then((module) => {
  // moduleの使用
});
3. package.json の "type": "module" を削除
/Users/sorao/Desktop/lgtm-factory/node_modules/lint-staged/package.json ファイルの "type": "module" を削除すると、そのスコープの .js ファイルは CommonJS モジュールとして扱われます。ただし、他の依存関係に影響を与える可能性があるため、注意が必要です。

4. 依存関係の設定を確認
lint-staged や husky の最新バージョンを確認し、可能であればアップデートすることで、この問題が修正される場合もあります。
```

1.を実施したがそれでもcommitができなかった...。

**Comment by kagomen** [comment]
@soraOrange 

自分もmainからブランチ切って同じようにやってみたんですが、同じ警告文出ませんでした。原因、なんだろう？

ただ、最後から2番目の行の`husky - pre-commit script failed (code 1)`というエラー文は出て、これが原因でコミットできなかったので、
soraさんがコミットできてない理由が`Warning: require() of ES modules is not supported.`ではなくて、そっちかな？と推測します👀

Warningは警告なので、エラーとはまた別かなと！

**Reply by ghost** [reply]
@kagomen  おおお、、、なんでだろう、、、
Nodeのversionとかそういうのでもなさそうなんですけどね;

**Comment by kazzyfrog** [comment]
@soraOrange 
ん、インストールした後に、package.jsonファイルにlint-stagedの設定を追加するまでは、OKですかね？
<img width="803" alt="スクリーンショット 2024-08-22 0 31 40" src="https://github.com/user-attachments/assets/968be592-41c8-4e4f-9f8a-6180b2926321">

package.jsonファイル
```json
{
  "lint-staged": {
    "*.{vue, ts, js}": "eslint --fix",
    "*.{js,jsx,ts,tsx,vue,css,html,md,json,cjs}": "prettier --write"
  }
}
```

**Reply by kazzyfrog** [reply]
自分の環境で検証したら、上記の記事の
- コマンドをコピペしてインストールした後に、
- package.jsonファイルにlint-stagedの設定を追加

をしたら、正常にコミットできました！


もし、これでもできなかったら、下記をご確認ください
- `git pull`で、最新のmainブランチの状態を、ローカルに反映させている

**Reply by ghost** [reply]
@kazzyfrog OKです！
その後に添付したエラーメッセージが表示されちゃうますね；

**Reply by ghost** [reply]
@kazzyfrog あ、メインブランチでやったらうまくいきました！！！
ただ、ブランチかえちゃうとうまくいかなくなっちゃいます；

**Reply by kazzyfrog** [reply]
@soraOrange 
そしたら、作業用ブランチの方も、最新の状態を取得する必要があるかも？です：
作業用ブランチの方で、
- まず現在加えている変更を一旦逃すことができます
  ```zsh
  git stash
  ```
- 次に、作業用ブランチも、最新の状態にします
  ```zsh
  git pull origin main
  ```
- 最後に、逃した変更を元に戻します
  ```zsh
  git stash pop
  ```

**Reply by ghost** [reply]
@kazzyfrog ありがとうございます！
↑で試してみたのですが、うまくいかないです....ね.....;

**Reply by kazzyfrog** [reply, marked as answer]
@soraOrange 
mainブランチと、作業ブランチに何か違う点はありますかね、、？

例えば、
 "lint-staged"の設定前後のカンマ忘れとか！
エラー見る限り、jsonの形式あたりのエラーな気がするんですよね。。！
```json
"devDependencies": {
    "@cloudflare/next-on-pages": "^1.13.2",
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "eslint": "^8",
    "eslint-config-next": "14.2.5",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-next-on-pages": "^1.13.2",
    "husky": "^9.1.5",
    "lint-staged": "^15.2.9",
    "postcss": "^8",
    "prettier": "^3.3.3",
    "prettier-plugin-tailwindcss": "^0.6.6",
    "tailwindcss": "^3.4.1",
    "typescript": "^5"
  },
  "lint-staged": {
    "*.{vue, ts, js}": "eslint --fix",
    "*.{js,jsx,ts,tsx,vue,css,html,md,json,cjs}": "prettier --write"
  }
```

**Reply by ghost** [reply]
@kazzyfrog いけましたーーーーーーーーーーーーーーーーーーー！！！！！！！！！！！！！！！
{}が多かったみたいでした........
失礼しました、、、、、、

**Reply by kazzyfrog** [reply]
@soraOrange 
良かったです！！！
確かに、記事の内容だと、少しわかりずらいですね！w

すぐに確認できるように、
[上記のコメント](https://github.com/lgtm-factory/lgtm-factory/discussions/86#discussioncomment-10409975)を、「Mark as answer」ボタンで、マークしてみてください！👍

マークすることで、このQAスレッドが、解決済みであることを表示できますー🙌

**Comment by ghost** [comment]
@kazzyfrog @kagomen 
ありがとうございます！！🐕

**Reply by kagomen** [reply]
良かったです！！
